docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+
+Once the update completes, the Machine controller will remove the name of the updater that has finished from the list of updaters and will start the next one. If the update fails, this will be reflected in the Machine status.
+
+If all updaters succeed, the controller will mark the Machine as up to date and remove the update plan from the Machine.
+
+### Infra Machine Template changes
+
+As mentioned before, the user experience to update in-place should be the exact same one as for rolling updates. This includes the need to rotate the Infra machine template. For providers that bundle the kubernetes components in some kind of image, this means that when upgrading kubernetes versions, a new image will be required.
+
+This might seem counter-intuitive, given the update will be made in-place, so there is no nee for a new image. However, not only this ensures the experience is the same as in rolling updates, but it also allows new Machines to be created for that MachineDeployment/CP in case a scale up or fallback to rolling is required.
+
+We leave up to the external updater implementers to decide how to deal with these changes. Some infra providers might have the ability to swap the image in an existing running machine, in which case they can offer a true in-place update for this field. For the ones that can't do this but want to allow changes that require a new image (like kubernetes updates), they should "ignore" the image field when processing the update, leaving the machine in a dirty state.
+
+We might explore the ability to represent this "dirty" state at the API level. We leave this for a future iteration of this feature.
+
+### Remediation
+
+Remediation can be used as the solution to recover machine when in-place update fails on a machine. The remediation process stays the same as today: the MachineHealthCheck controller monitors machine health status and marks it to be remediated based on pre-configured rules, then ControlPlane/MachineDeployment replaces the machine or call external remediation.
+
+However, in-place updates might cause Nodes to become unhealthy while the update is in progress. In addition, an in-place update might take more (or less) time than a fresh machine creation. Hence, in order to successfully use MHC to remediate in-place updated Machines, we require:
+* A mechanism to identify if a Machine is being updated. We will surface this in the Machine status. API details will be added later.
+* A way to define different rules for Machines on-going an update. This might involve new fields in the MHC object. We will decouple these API changes from this proposal. For the first implementation of in-place updates, we might decide to just disable remediation for Machines that are on-going an update.
+
+
+### API Changes
```

Comment:
It'd be great to start the proposal with an example in how we envision the end state to look like, from defining the state and provide an in depth example with KCP, kubeadm bootstrap provider, and an example infra provider (like AWS or similar)

Reply:
Are you suggesting we do this before we define the API changes? or as part of that work?

We purposefully left the API design for later so we can focus the conversation on the core ideas and high level flow and make sure we are aligned there first.

Reply:
Without seeing the API changes we're proposing it's generally hard to grasp the high level concept. I would like to see it from a user/operator perspective:

- How will we setup this feature in yaml?
- What are the required pieces that we need to install?
- Are there any assumptions we're making?